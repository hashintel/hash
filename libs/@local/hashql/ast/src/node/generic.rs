use hashql_core::{span::SpanId, symbol::Ident};

use super::{id::NodeId, r#type::Type};
use crate::heap;

/// A concrete type argument for a generic type.
///
/// Represents a specific type used to instantiate a generic type parameter.
/// In a generic type instantiation like `Vec<String>`, `String` would be
/// represented as a [`GenericArgument`].
///
/// # Examples
///
/// In a type reference like `HashMap<K, V>`, `K` and `V` are generic arguments.
/// In an instantiated type like `List<Int>`, `Int` is a generic argument.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct GenericArgument {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
}

/// A generic type parameter declaration.
///
/// Represents a type parameter in a generic type or function declaration.
/// Each parameter has a name and an optional type bound that constrains
/// what types can be used as arguments for this parameter.
///
/// # Examples
///
/// In a declaration like `fn<T>(x: T): T`, `T` is a generic parameter.
/// In a declaration like `fn<T: Number>(x: T): T`, `T` is a generic parameter
/// with a bound of `Number`.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct GenericParam<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
    pub bound: Option<heap::Box<'heap, Type<'heap>>>,
}

/// A collection of generic parameters for a type or function.
///
/// Represents the complete set of type parameters in a generic declaration,
/// such as the `<T, U>` in `fn<T, U>(x: T, y: U): U`.
///
/// # Examples
///
/// For a function like `fn<T, U: Comparable>(x: T, y: U): Bool`:
/// - The `Generics` would contain two `GenericParam` entries
/// - One for `T` with no bound
/// - One for `U` with a bound of `Comparable`
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Generics<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub params: heap::Box<'heap, [GenericParam<'heap>]>,
}
