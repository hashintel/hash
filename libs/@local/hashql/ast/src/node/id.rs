use hashql_core::id;

id::newtype!(
    /// A unique identifier for nodes in the HashQL Abstract Syntax Tree.
    ///
    /// Each node in the AST has a unique `NodeId` that can be used to track the node
    /// through various processing stages. These IDs are particularly important for
    /// error reporting, cross-referencing between nodes, and maintaining node identity
    /// during transformations.
    ///
    /// The value space is restricted to 0..=0xFFFF_FF00, reserving the last 256 for niches.
    /// As real pattern types are an experimental feature in Rust, these can currently only be
    /// used by directly modifying and accessing the `NodeId`'s internal value.
    ///
    /// # Implementation Note
    ///
    /// We're currently liberal in the use of `NodeId`s. In the future, we might want to
    /// revisit this to see if we can use fewer of them or optimize their allocation.
    pub struct NodeId(u32 is 0..=0xFFFF_FF00)
);

impl NodeId {
    /// A placeholder ID used during initial parsing.
    ///
    /// When parsing, we initially give all AST nodes this placeholder ID. After resolving special
    /// forms, the nodes are renumbered to have small indices.
    pub const PLACEHOLDER: Self = <Self as id::Id>::MAX;
}
