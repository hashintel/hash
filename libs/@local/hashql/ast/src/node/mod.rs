//! Core Abstract Syntax Tree node types for HashQL.
//!
//! This module defines the fundamental building blocks of the HashQL Abstract
//! Syntax Tree (AST). The AST represents the structure of a HashQL program after
//! parsing, capturing all the language constructs in a hierarchical tree structure.
//!
//! ## Module Organization
//!
//! The node module is organized into several submodules:
//!
//! - [`expr`]: Expression nodes representing computations that produce values
//! - [`generic`]: Generic type parameter and argument definitions
//! - [`id`]: Node identifier types and utilities
//! - [`path`]: Path expressions for variable references and qualified names
//! - [`type`]: Type system representations including struct, tuple, and other types
//!
//! ## Memory Management
//!
//! All AST nodes use arena allocation through the `'heap` lifetime parameter,
//! which references the memory region where nodes are allocated. This approach
//! improves performance by minimizing allocations and simplifies memory management
//! during parsing and analysis.
//!
//! ## Node Identification
//!
//! Each node in the AST has a unique [`id::NodeId`] that can be used to track
//! the node through various processing stages. Nodes also include a span identifier that points to
//! the source location of the node.
//!
//! ## AST Structure
//!
//! The AST is structured as a tree with expressions ([`expr::Expr`]) at the core.
//! Other node types like types ([`type::Type`]) and paths ([`path::Path`]) are
//! used as components within expressions or on their own for declarations.
//!
//! The entire tree represents a complete HashQL program or query, ready for
//! analysis, transformation, and evaluation.
pub mod expr;
pub mod generic;
pub mod id;
pub mod path;
pub mod r#type;
