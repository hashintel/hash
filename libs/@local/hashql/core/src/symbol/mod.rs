//! Symbol representation and manipulation for HashQL.
//!
//! This module defines dedicated types for representing symbols in HashQL's compilation process.
//! Symbols are string-like values that appear in source code, such as identifiers, keywords,
//! and literals. They are deliberately kept separate from the syntax tree as they are used through
//! the different stages of compilation.
//!
//! The module provides:
//!
//! - [`Symbol`]: An opaque wrapper around string data that enables efficient storage and comparison
//! - [`Ident`]: A named identifier with source location and categorization
//! - [`IdentKind`]: Classification of different identifier types in HashQL
//!
//! ## Design Philosophy
//!
//! The [`Symbol`] type is designed as an opaque wrapper around its internal string storage.
//! This encapsulation enables future optimizations such as string interning (either through
//! the `string_interner` crate or a custom implementation) without requiring API changes.

pub mod sym;

use core::{
    borrow::Borrow,
    cmp::Ordering,
    fmt::{self, Display, Formatter},
    hash::{Hash, Hasher},
    ptr,
};

use crate::span::SpanId;

/// A string-like value used throughout the HashQL compiler.
///
/// Symbols represent string data that appears in source code and persists throughout
/// compilation, they are read-only and immutable.
///
/// This type is deliberately opaque to hide its internal representation,
/// allowing for future optimizations like string interning without changing
/// the public API. Symbols are designed to be efficient for long-lived objects
/// that are frequently compared, hashed, and referenced during compilation.
///
/// The caller must ensure that the string is unique and interned. The types correctness requires
/// relies on these *but it does not enforce it*.
#[derive(Debug, Copy, Clone)]
pub struct Symbol<'heap>(&'heap str);

impl<'heap> Symbol<'heap> {
    /// Creates a new interned symbol from a string slice.
    ///
    /// The caller must ensure that the string is unique and interned.
    pub(crate) const fn new_unchecked(string: &'heap str) -> Self {
        Self(string)
    }

    #[must_use]
    pub const fn as_str(&self) -> &str {
        self.0
    }

    /// Returns the string representation of the symbol.
    ///
    /// Unlike [`Self::as_str`], this method provides access for the lifetime of the interner
    /// instead of the symbol itself, somewhat circumventing the protections given to the symbol
    /// itself. Any unwrapped type should be considered no longer unique and interned.
    #[must_use]
    pub const fn unwrap(&self) -> &'heap str {
        self.0
    }

    #[must_use]
    pub const fn as_bytes(&self) -> &[u8] {
        self.0.as_bytes()
    }
}

impl AsRef<Self> for Symbol<'_> {
    #[inline]
    fn as_ref(&self) -> &Self {
        self
    }
}

impl PartialEq for Symbol<'_> {
    fn eq(&self, other: &Self) -> bool {
        // Pointer equality implies string equality (due to the unique contents assumption)
        ptr::eq(self.0, other.0)
    }
}

impl Eq for Symbol<'_> {}

impl PartialOrd for Symbol<'_> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Symbol<'_> {
    fn cmp(&self, other: &Self) -> Ordering {
        // Pointer equality implies string equality (due to the unique contents assumption), but if
        // not the same the contents must be compared.
        if self == other {
            Ordering::Equal
        } else {
            self.0.cmp(other.0)
        }
    }
}

impl Hash for Symbol<'_> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Pointer hashing is sufficient (due to the unique contents assumption)
        ptr::hash(self.0, state);
    }
}

impl Borrow<str> for Symbol<'_> {
    fn borrow(&self) -> &str {
        self.0
    }
}

impl AsRef<str> for Symbol<'_> {
    fn as_ref(&self) -> &str {
        self.0
    }
}

impl Display for Symbol<'_> {
    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
        Display::fmt(self.0, fmt)
    }
}

/// The classification of an identifier in HashQL.
///
/// HashQL supports different categories of identifiers, each with distinct
/// syntactic rules and semantic meanings in the language.
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum IdentKind {
    /// A lexical identifier following standard programming language naming rules.
    ///
    /// # Syntax
    ///
    /// Lexical identifiers in HashQL follow Rust's identifier parsing rules:
    /// - Begin with either an `XID_Start` Unicode character or an underscore (`_`)
    /// - Continue with zero or more `XID_Continue` Unicode characters
    /// - Cannot be a single underscore character (`_`)
    /// - Uppercase identifiers are used for types, while lowercase identifiers are used for values,
    ///   identifiers starting with an underscore (`_`) are considered lowercase.
    ///
    /// In formal notation:
    /// ```text
    /// IDENTIFIER_LEXICAL:
    ///     XID_Start XID_Continue*
    ///     | '_' XID_Continue+
    /// ```
    ///
    /// > Note: Zero-width characters are not allowed.
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// counter
    /// _identifier
    /// Москва        // Non-ASCII identifiers are supported
    /// 東京           // As are other scripts
    /// ```
    Lexical,

    /// A symbolic identifier consisting of operator-like characters.
    ///
    /// # Syntax
    ///
    /// Symbol identifiers can consist of ASCII operators or Unicode symbols/punctuation:
    ///
    /// ```text
    /// ASCII_SYMBOL:
    ///       '!' | '#' | '$' | '%' | '&'
    ///     | '*' | '+' | '.' | '/' | '<'
    ///     | '=' | '>' | '?' | '@' | '\'
    ///     | '^' | '|' | '-' | '~'
    ///
    /// RESERVED_SYMBOL:
    ///     ':'
    ///
    /// ; Excludes any symbols that are in the ASCII range
    /// UNICODE_SYMBOL:
    ///     Symbol
    ///     | Punctuation
    ///
    /// SYMBOL:
    ///     ASCII_SYMBOL
    ///     | UNICODE_SYMBOL
    ///
    /// IDENTIFIER_SYMBOL:
    ///     SYMBOL+
    ///     | '`' SYMBOL+ '`'
    /// ```
    ///
    /// > Note: `:` is reserved, and cannot be used as a symbol identifier.
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// <
    /// +
    /// ->
    /// ≥
    /// →
    /// `*`
    /// ```
    Symbol,

    /// An identifier representing a URL for external resource references.
    ///
    /// # Syntax
    ///
    /// Base URL identifiers must be valid `http` or `https` URLs that end with a forward slash
    /// (`/`), and must be enclosed within backticks:
    ///
    /// ```text
    /// IDENTIFIER_BASE_URL:
    ///     '`' HTTP_URL_STRING_ENDING_WITH_SLASH '`'
    /// ```
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// `https://blockprotocol.org/@blockprotocol/types/property-type/address/`
    /// `https://example.com/resources/schema/`
    /// `http://localhost:8080/api/v1/`
    /// ```
    BaseUrl,
}

/// A named identifier in HashQL with source location information.
///
/// An `Ident` combines a [`Symbol`] (the textual content of the identifier)
/// with a [`SpanId`] (its location in the source code) and a [`IdentKind`]
/// (its syntactic classification). Identifiers are used throughout the compiler
/// to represent variables, functions, types, and other named entities.
///
/// # Examples
///
/// In HashQL source code:
///
/// ```text
/// let counter = 0;
///     ^^^^^^^
/// ```
///
/// This identifier would be represented as an `Ident` with:
/// - name: `Symbol("counter")`
/// - kind: `IdentKind::Lexical`
/// - span: `5..12`
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Ident<'heap> {
    pub span: SpanId,

    pub value: Symbol<'heap>,
    pub kind: IdentKind,
}

#[cfg(test)]
impl<'heap> Ident<'heap> {
    pub(crate) const fn synthetic(value: Symbol<'heap>) -> Self {
        Self {
            span: SpanId::SYNTHETIC,
            value,
            kind: IdentKind::Lexical,
        }
    }
}

impl AsRef<str> for Ident<'_> {
    fn as_ref(&self) -> &str {
        self.value.as_str()
    }
}

impl Display for Ident<'_> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        Display::fmt(&self.value.0, fmt)
    }
}
