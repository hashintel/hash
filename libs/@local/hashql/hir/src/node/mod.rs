pub mod access;
pub mod branch;
pub mod call;
pub mod closure;
pub mod data;
pub mod graph;
pub mod input;
pub mod kind;
pub mod r#let;
pub mod operation;
pub mod variable;

use hashql_core::{id, span::SpanId};

use self::kind::NodeKind;

id::newtype!(
    /// A unique identifier for nodes in the HashQL High-Level Intermediate Representation (HIR).
    ///
    /// Each node in the HIR has a unique `NodeId` that can be used to track the node
    /// through various processing stages. These IDs are particularly important for
    /// error reporting, cross-referencing between nodes, and maintaining node identity
    /// during transformations.
    ///
    /// The value space is restricted to 0..=0xFFFF_FF00, reserving the last 256 for niches.
    /// As real pattern types are an experimental feature in Rust, these can currently only be
    /// used by directly modifying and accessing the `NodeId`'s internal value.
    pub struct HirId(u32 is 0..=0xFFFF_FF00)
);

impl HirId {
    /// A placeholder ID used during initial parsing.
    ///
    /// When parsing, we initially give all AST nodes this placeholder ID. After resolving special
    /// forms, the nodes are renumbered to have small indices.
    pub const PLACEHOLDER: Self = <Self as id::Id>::MAX;
}

/// A node in the HashQL High-Level Intermediate Representation (HIR).
///
/// The HIR is an optimized, more refined representation of the program derived from the AST.
/// While the AST closely mirrors the source syntax, the HIR represents a more semantically
/// meaningful structure that's better suited for type checking, optimization, and code generation.
///
/// Key differences between HIR and AST:
/// - Special forms (like `let`, `if`, etc.) are fully expanded in the HIR
/// - Syntactic sugar is desugared and normalized
/// - The HIR uses interned references for better memory efficiency
/// - Nodes are more compact and optimized for compiler operations
///
/// Each HIR node has:
/// - A unique identifier for tracking through compilation phases
/// - A span linking to its origin in source code for error reporting
/// - A specific kind that determines the node's semantics and structure
///
/// The HIR is designed to be immutable after construction and uses arena allocation
/// through the `'heap` lifetime for efficient memory management.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct Node<'heap> {
    pub id: HirId,
    pub span: SpanId,

    pub kind: NodeKind<'heap>,
}
