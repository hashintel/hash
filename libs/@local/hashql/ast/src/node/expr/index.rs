use hashql_core::{heap, span::SpanId};

use super::Expr;
use crate::node::id::NodeId;

/// An indexing expression in the HashQL Abstract Syntax Tree.
///
/// Represents accessing an element from a collection by its index or key.
/// Indexing works with various collection types including lists, tuples,
/// dictionaries, structs, and other indexed data structures.
///
/// The index expression can be any expression that evaluates to a valid
/// index for the collection type - typically an integer for lists or a value of appropriate key
/// type for dictionaries.
///
/// Tuples and structs can be indexed using this expression, but only if the type system can prove
/// that the index is within bounds.
///
/// # Examples
///
/// ## J-Expr
///
/// ```json
/// ["[]", "items", 0]
/// ["items[0]"]
/// ["matrix[i][j]"]
/// ```
///
/// ## Documentation Format
///
/// ```text
/// items[0]
/// matrix[i][j]
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct IndexExpr<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub value: heap::Box<'heap, Expr<'heap>>,
    pub index: heap::Box<'heap, Expr<'heap>>,
}
