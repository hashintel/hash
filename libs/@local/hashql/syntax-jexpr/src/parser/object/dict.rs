use hashql_ast::node::{
    expr::{
        DictExpr, Expr, ExprKind, LiteralExpr,
        dict::DictEntry,
        literal::{LiteralKind, StringLiteral},
    },
    id::NodeId,
};
use hashql_core::symbol::Symbol;
use text_size::TextRange;

use super::{
    ObjectState, State,
    error::{
        dict_entry_expected_array, dict_entry_too_few_items, dict_entry_too_many_items,
        dict_expected_format,
    },
    r#type::{TypeNode, handle_typed},
    visit::{Key, visit_object},
};
use crate::{
    ParserState,
    error::ResultExt as _,
    lexer::{syntax_kind::SyntaxKind, syntax_kind_set::SyntaxKindSet, token::Token},
    parser::{
        array::visit::visit_array, error::ParserDiagnostic, expr::parse_expr, state::Expected,
    },
};

pub(crate) struct DictNode<'heap> {
    key_span: TextRange,

    expr: DictExpr<'heap>,
    r#type: Option<TypeNode<'heap>>,
}

impl<'heap> DictNode<'heap> {
    pub(crate) fn parse(
        state: &mut ParserState<'heap, '_>,
        key: &Key<'_>,
    ) -> Result<Self, ParserDiagnostic> {
        let expr = parse_dict(state)?;

        Ok(Self {
            key_span: key.span,
            expr,
            r#type: None,
        })
    }

    pub(crate) fn with_type(mut self, r#type: TypeNode<'heap>) -> Self {
        self.r#type = Some(r#type);
        self
    }
}

impl<'heap> State<'heap> for DictNode<'heap> {
    fn handle(
        mut self,
        state: &mut ParserState<'heap, '_>,
        key: Key<'_>,
    ) -> Result<ObjectState<'heap>, ParserDiagnostic> {
        handle_typed("#dict", self.key_span, &mut self.r#type, state, &key)?;
        Ok(ObjectState::Dict(self))
    }

    fn build(
        mut self,
        state: &mut ParserState<'heap, '_>,
        span: TextRange,
    ) -> Result<Expr<'heap>, ParserDiagnostic> {
        self.expr.r#type = TypeNode::finish(self.r#type, state);

        Ok(Expr {
            id: NodeId::PLACEHOLDER,
            span: state.insert_range(span),
            kind: ExprKind::Dict(self.expr),
        })
    }
}

fn parse_dict_object<'heap, 'source>(
    state: &mut ParserState<'heap, 'source>,
    token: Token<'source>,
) -> Result<DictExpr<'heap>, ParserDiagnostic> {
    let mut entries = Vec::new();

    let span = visit_object(state, token, |state, key| {
        let key_span = key.span;

        let key_span_id = state.insert_range(key.span);
        let key = Expr {
            id: NodeId::PLACEHOLDER,
            span: key_span_id,
            kind: ExprKind::Literal(LiteralExpr {
                id: NodeId::PLACEHOLDER,
                span: key_span_id,
                kind: LiteralKind::String(StringLiteral {
                    value: Symbol::new(key.value),
                }),
                r#type: None,
            }),
        };

        let value = parse_expr(state)?;
        let value_span = state.current_span();

        let entry_span = key_span.cover(value_span);

        entries.push(DictEntry {
            id: NodeId::PLACEHOLDER,
            span: state.insert_range(entry_span),
            key: state.heap().boxed(key),
            value: state.heap().boxed(value),
        });

        Ok(())
    })?;

    Ok(DictExpr {
        id: NodeId::PLACEHOLDER,
        span: state.insert_range(span),
        entries: state.heap().boxed_slice(entries),
        r#type: None,
    })
}

fn parse_dict_array<'heap, 'source>(
    state: &mut ParserState<'heap, 'source>,
    token: Token<'source>,
) -> Result<DictExpr<'heap>, ParserDiagnostic> {
    let mut entries = Vec::new();

    let span = visit_array(state, token, |state| {
        let mut key = None;
        let mut value = None;
        let mut excess = Vec::new();

        // We're parsing everything here, so that we're able to improve the error message
        let token = state
            .advance(Expected::hint(SyntaxKind::LBracket))
            .change_category(From::from)?;

        if token.kind.syntax() != SyntaxKind::LBracket {
            return Err(dict_entry_expected_array(
                state.insert_range(token.span),
                token.kind.syntax(),
            )
            .map_category(From::from));
        }

        let span = visit_array(state, token, |state| {
            if key.is_some() && value.is_some() {
                // We just parse, and then report the issue later
                // This way we're able to tell the user how many entries were skipped
                let expr = parse_expr(state)?;
                excess.push(expr.span);

                return Ok(());
            }

            let expr = parse_expr(state)?;

            if key.is_none() {
                key = Some(expr);
            } else {
                value = Some(expr);
            }

            Ok(())
        })?;

        if !excess.is_empty() {
            return Err(dict_entry_too_many_items(&excess).map_category(From::from));
        }

        let found = usize::from(key.is_some()) + usize::from(value.is_some());
        let Some((key, value)) = Option::zip(key, value) else {
            return Err(
                dict_entry_too_few_items(state.insert_range(span), found).map_category(From::from)
            );
        };

        entries.push(DictEntry {
            id: NodeId::PLACEHOLDER,
            span: state.insert_range(span),
            key: state.heap().boxed(key),
            value: state.heap().boxed(value),
        });

        Ok(())
    })?;

    Ok(DictExpr {
        id: NodeId::PLACEHOLDER,
        span: state.insert_range(span),
        entries: state.heap().boxed_slice(entries),
        r#type: None,
    })
}

fn parse_dict<'heap>(
    state: &mut ParserState<'heap, '_>,
) -> Result<DictExpr<'heap>, ParserDiagnostic> {
    // We're parsing everything here, so that we're able to improve the error message
    let token = state
        .advance(Expected::hint(SyntaxKindSet::from_slice(&[
            SyntaxKind::LBrace,
            SyntaxKind::LBracket,
        ])))
        .change_category(From::from)?;

    let is_object = match token.kind.syntax() {
        SyntaxKind::LBrace => true,
        SyntaxKind::LBracket => false,
        kind => {
            let span = state.insert_range(token.span);

            return Err(dict_expected_format(span, kind)).change_category(From::from)?;
        }
    };

    if is_object {
        parse_dict_object(state, token)
    } else {
        parse_dict_array(state, token)
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_snapshot, with_settings};

    use crate::{
        lexer::syntax_kind::SyntaxKind,
        parser::{object::parse_object, test::bind_parser},
    };

    // Create a parser binding that will handle objects starting with '{'
    bind_parser!(fn parse_object_expr(parse_object, SyntaxKind::LBrace));

    #[test]
    fn parse_empty_dict_object_format() {
        // Empty dict with object format
        let result = parse_object_expr(r##"{"#dict": {}}"##)
            .expect("should parse empty dict in object format");

        with_settings!({
            description => "Parses an empty dict using object format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_empty_dict_array_format() {
        // Empty dict with array format
        let result = parse_object_expr(r##"{"#dict": []}"##)
            .expect("should parse empty dict in array format");

        with_settings!({
            description => "Parses an empty dict using array format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_incomplete() {
        // Empty dict with object format
        let result =
            parse_object_expr(r##"{"#dict": "##).expect_err("should not parse incomplete dict");

        with_settings!({
            description => "Parses with a sudden EOF"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.diagnostic, &result.input);
        });
    }

    #[test]
    fn parse_dict_object_format() {
        // Dict with entries in object format
        let result = parse_object_expr(
            r##"{"#dict": {
            "key1": {"#literal": "value1"},
            "key2": {"#literal": 42}
        }}"##,
        )
        .expect("should parse dict in object format");

        with_settings!({
            description => "Parses a dict with entries using object format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_array_format() {
        // Dict with entries in array format
        let result = parse_object_expr(
            r##"{"#dict": [
            [{"#literal": "key1"}, {"#literal": "value1"}],
            [{"#literal": "key2"}, {"#literal": 42}]
        ]}"##,
        )
        .expect("should parse dict in array format");

        with_settings!({
            description => "Parses a dict with entries using array format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_with_complex_values() {
        // Dict with complex values in object format
        let result = parse_object_expr(
            r##"{"#dict": {
            "simple": {"#literal": "value"},
            "nested": {"#struct": {"name": {"#literal": "Alice"}}},
            "list": {"#list": [{"#literal": 1}, {"#literal": 2}]},
            "function": ["add", {"#literal": 1}, {"#literal": 2}]
        }}"##,
        )
        .expect("should parse dict with complex values");

        with_settings!({
            description => "Parses a dict with complex values using object format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_array_format_with_complex_values() {
        // Dict with complex values in array format
        let result = parse_object_expr(
            r##"{"#dict": [
            [{"#literal": "key1"}, {"#struct": {"name": {"#literal": "Bob"}}}],
            [{"#literal": "key2"}, {"#list": [{"#literal": 1}, {"#literal": 2}]}]
        ]}"##,
        )
        .expect("should parse dict with complex values in array format");

        with_settings!({
            description => "Parses a dict with complex values using array format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_with_type() {
        // Dict with type annotation
        let result = parse_object_expr(
            r##"{"#dict": {
            "key": {"#literal": "value"}
            }, "#type": "Dict<String, String>"}"##,
        )
        .expect("should parse dict with type");

        with_settings!({
            description => "Parses a dict with type annotation"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_with_complex_type() {
        // Dict with complex type annotation
        let result = parse_object_expr(
            r##"{"#dict": {
            "key": {"#literal": "value"}
        }, "#type": "Dict<String, `?`>"}"##,
        )
        .expect("should parse dict with complex type");

        with_settings!({
            description => "Parses a dict with complex type annotation"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_dict_array_format_with_too_few_items() {
        // Array format dict entry with too few items
        let error = parse_object_expr(
            r##"{"#dict": [
            [{"#literal": "key1"}]
        ]}"##,
        )
        .expect_err("should fail with too few items");

        with_settings!({
            description => "Rejects dict entry with too few items in array format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_dict_array_format_with_too_many_items() {
        // Array format dict entry with too many items
        let error = parse_object_expr(
            r##"{"#dict": [
            [{"#literal": "key1"}, {"#literal": "value1"}, {"#literal": "extra"}]
        ]}"##,
        )
        .expect_err("should fail with too many items");

        with_settings!({
            description => "Rejects dict entry with too many items in array format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_dict_invalid_array_entry() {
        // Array format with invalid entry (not an array)
        let error = parse_object_expr(
            r##"{"#dict": [
            {"not": "an-array"}
        ]}"##,
        )
        .expect_err("should fail with invalid array entry");

        with_settings!({
            description => "Rejects invalid entry format in array-format dict"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_invalid_dict_format() {
        // Invalid dict value (neither object nor array)
        let error = parse_object_expr(r##"{"#dict": "invalid"}"##)
            .expect_err("should fail with invalid dict format");

        with_settings!({
            description => "Rejects invalid dict format (not object or array)"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_duplicate_dict_key() {
        // Duplicate #dict keys
        let error = parse_object_expr(r##"{"#dict": {}, "#dict": {}}"##)
            .expect_err("should fail with duplicate #dict key");

        with_settings!({
            description => "Rejects duplicate #dict keys"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_duplicate_type_key() {
        // Duplicate #type keys
        let error = parse_object_expr(
            r##"{"#dict": {}, "#type": "Dict<String, Int>", "#type": "Dict<String, Boolean>"}"##,
        )
        .expect_err("should fail with duplicate #type key");

        with_settings!({
            description => "Rejects duplicate #type keys in dict"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_unknown_key_in_dict() {
        // Unknown key in dict object
        let error = parse_object_expr(r##"{"#dict": {}, "unknown": {"#literal": "value"}}"##)
            .expect_err("should fail with unknown key");

        with_settings!({
            description => "Rejects unknown keys in dict objects"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_dict_with_complex_keys() {
        // Dict with complex expressions as keys in array format
        let result = parse_object_expr(
            r##"{"#dict": [
            [["concat", {"#literal": "pre"}, {"#literal": "fix"}], {"#literal": "value"}],
            [{"#struct": {"keypart": {"#literal": "complex"}}}, {"#literal": "value2"}]
        ]}"##,
        )
        .expect("should parse dict with complex keys in array format");

        with_settings!({
            description => "Parses a dict with complex expressions as keys using array format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_nested_dicts() {
        // Nested dicts in object format
        let result = parse_object_expr(
            r##"{"#dict": {
            "outer1": {"#dict": {
                "inner1": {"#literal": "value1"},
                "inner2": {"#literal": "value2"}
            }},
            "outer2": {"#dict": {
                "inner3": {"#literal": "value3"}
            }}
        }}"##,
        )
        .expect("should parse nested dicts");

        with_settings!({
            description => "Parses nested dicts using object format"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }
}
