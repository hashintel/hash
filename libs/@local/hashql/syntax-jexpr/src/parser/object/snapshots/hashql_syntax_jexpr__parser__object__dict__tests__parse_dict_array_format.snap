---
source: libs/@local/hashql/syntax-jexpr/src/parser/object/dict.rs
description: Parses a dict with entries using array format
expression: "{\"#dict\": [\n            [{\"#literal\": \"key1\"}, {\"#literal\": \"value1\"}],\n            [{\"#literal\": \"key2\"}, {\"#literal\": 42}]\n        ]}"
---
Expr#4294967040@11
  ExprKind (Dict)
    DictExpr#4294967040@10
      DictEntry#4294967040@4
        Expr#4294967040@1
          ExprKind (Literal)
            LiteralExpr#4294967040@0
              LiteralKind (String)
                StringLiteral#4294967040@0 (key1)
        Expr#4294967040@3
          ExprKind (Literal)
            LiteralExpr#4294967040@2
              LiteralKind (String)
                StringLiteral#4294967040@2 (value1)
      DictEntry#4294967040@9
        Expr#4294967040@6
          ExprKind (Literal)
            LiteralExpr#4294967040@5
              LiteralKind (String)
                StringLiteral#4294967040@5 (key2)
        Expr#4294967040@8
          ExprKind (Literal)
            LiteralExpr#4294967040@7
              LiteralKind (Integer)
                IntegerLiteral#4294967040@7 (42)
